### DISCUSSION TITLE
[LUA] detect difference between a game start and restart?

### DISCUSSION THREAD

**Discussion by Jakobud** [mamedev/discussions#161] [Q&A] [answered]
From what I understand you would typically use `emu.add_machine_reset_notifier(callback)` to make your LUA plugin do something when a game starts or is restarted.

However I recently came across a MAME rom: `zwackery`

I noticed that this game has a startup procedure like most games (checking roms, sound, etc). It lasts for maybe 10+ seconds or so.

But if you hit reset (F3) for this game, it restarts the game but the startup sequence is shortened to only 1-2 seconds. It appears to skip a lot of the sequence from when you start the game initially.

Are these two different events in LUA? the `emu.add_machine_reset_notifier(callback)` seems to happen for both cases, but there are clearly different things happening here.

There are probably other games with this behavior but this is the first I've noticed.

**Comment by holub** [comment]
Not sure about precise names but initial startup is hard reset which calls "start+reset" and can be reinitiated with Shift+F3
Soft reset is just "reset" which is F3

**Reply by Jakobud** [reply]
Thank you but I'm asking about the LUA API.

**Reply by holub** [reply]
https://github.com/mamedev/mame/blob/e3907cb6977d19d973b30f08bfaa0501b5ccdb79/src/frontend/mame/luaengine.cpp#L778

**Reply by Jakobud** [reply]
Okay so `emu.register_start(callback)` is deprecated and has been replaced with `emu.add_machine_reset_notifier(callback)`. I'm asking specifically about if there is a different notifier for when a game starts vs when it resets and why the reset notifier is triggered for both starts and resets when they appear to do different things in some games.

**Comment by rb6502** [comment]
MAME has a defined order of operations.  When you start up an emulated machine, it first calls "start" on everything, and then it calls "reset" on everything.  This is pretty much how all digital hardware works if you think of the "start" as turning the power on.  First you apply power, and then some circuit automatically resets (at least) the CPU so that it can start running in a known state.

**Reply by Jakobud** [reply]
Okay I think I understand that. Does the LUA api differentiate between the 2 somehow?

Try running the rom `zwackery`. Wait for it to go through the boot sequence. Then pretty reset (F3) and see the difference. Do the LUA notifiers differentiate between these 2 behaviors?

**Comment by cuavas** [comment, marked as answer]
There currently isn’t a way to get a callback on start.  `emu.register_start` was named misleadingly – it was called on reset.  It’s something I’d like to add, but I don’t want to go changing the callbacks without thinking it through properly – I don’t want to create churn, particularly when it might affect core classes like the running machine and the externally visible Lua API.

**Reply by Jakobud** [reply]
Thank you for the insight and brilliant work you do!

**Comment by galibert** [comment]
Perhaps an extra bool parameter to the reset callback?  Lua is very
permissive there.

**Reply by Jakobud** [reply]
Can you elaborate on what you mean by this? 

I was also thinking about just setting a bool variable in the plugin that gets toggled off when the game exits. I would set it true when the game starts and then if I reset (F3) then it would still be true and know that it's a soft reset. However that wouldn't help with the hard reset as it would still be true I think. I'll experiment with this.

**Comment by Jakobud** [comment]
Okay I found a workaround to this.

`local running = false`

Then at start of my `emu.add_machine_reset_notifier` function:

```
if not running then
  print("start or hard reset detected")
  running = true
else
  print("soft reset detected")
end
```

And then in my `emu.add_machine_stop_notifier` function:

`running = false`

This appears to accurately differentiate between a start/hard reset and a soft reset.
